pkg/apiserver/registry/networkpolicy/clustergroupmember/rest.go

```diff
@@ -70,9 +81,52 @@ func (r *REST) Get(ctx context.Context, name string, options *metav1.GetOptions)
 		memberList.EffectiveMembers = effectiveMembers
 	}
 	memberList.Name = name
-	return memberList, nil
+	memberList.TotalMembers = int64(len(memberList.EffectiveMembers))
+	err = paginateMemberList(memberList, pageInfo)
+	return memberList, err
+}
+
+func (r *REST) NewGetOptions() (runtime.Object, bool, string) {
+	return &controlplane.PaginationGetOptions{}, false, ""
 }
 
 func (r *REST) NamespaceScoped() bool {
 	return false
 }
+
+// paginateMemberList returns paginated results if rational options are provided.
+// Paginated results are continuous only when there is no member change across multiple calls.
+// Returns full memberList if page number or page limit are given incorrectly.
+func paginateMemberList(memberList *controlplane.ClusterGroupMembers, pageInfo paginationInfo) error {
+	if memberList.TotalMembers == 0 {
+		memberList.TotalPages, memberList.CurrentPage = 0, 0
+		return nil
+	}
+	if pageInfo.pageLimit <= 0 {
+		memberList.TotalPages, memberList.CurrentPage = 1, 1
+		return nil
+	}
+	totalPages := (int64(len(memberList.EffectiveMembers)) + pageInfo.pageLimit - 1) / pageInfo.pageLimit
+	if totalPages >= pageInfo.pageNumber && pageInfo.pageNumber > 0 {
+		// Sort members based on name of ee/pod to realize consistent pagination support.
+		sort.SliceStable(memberList.EffectiveMembers, func(i, j int) bool {
+			if memberList.EffectiveMembers[i].Pod != nil && memberList.EffectiveMembers[j].Pod != nil {
+				return memberList.EffectiveMembers[i].Pod.Name < memberList.EffectiveMembers[j].Pod.Name
+			} else if memberList.EffectiveMembers[i].ExternalEntity != nil && memberList.EffectiveMembers[j].ExternalEntity != nil {
+				return memberList.EffectiveMembers[i].ExternalEntity.Name < memberList.EffectiveMembers[j].ExternalEntity.Name
+			} else {
+				return true
+			}
+		})
+		beginMember := (pageInfo.pageNumber - 1) * pageInfo.pageLimit
+		memberList.EffectiveMembers = memberList.EffectiveMembers[beginMember:]
+		if pageInfo.pageLimit < int64(len(memberList.EffectiveMembers)) {
+			memberList.EffectiveMembers = memberList.EffectiveMembers[:pageInfo.pageLimit]
+		}
+	} else {
+		pageInfo.pageNumber = 0 // set to 0 in case given page number exceeds total pages
```

Comment:
If I understand correctly if a user requires a page whose number is higher than the total number of pages they will get back the first page of results? Why not an empty page?

Reply:
They will get back the entire result of members, as if pagination is not requested. In the response, `totalPages` is the number of pages based on user's correct `limit`, but `pageNumber` is set to `0` to indicate that no specific page is returned.
Inspired by comments, I checked K8s pagination, in the case of `limit<=0`, pagination is ignored.

Reply:
I feel this may cause client a bit complex to handle this situation. For example, if some members are removed between two page queries, leading to the second query get empty result, and we return the entire result of members, it would waste the bandwidth spent on transfering the whole data and client has to add a page number check to distinguish whether this is a legal response before displaying it.

Reply:
That use case makes total sense, I'll change it to empty result, thanks!